### ISSUE TITLE
Add ability to change ntsOId and xNId

### ISSUE THREAD

**Issue by otterdahl** [rsmp-nordic/rsmp_validator#59]
The ability to configure ntsOId and xNId is useful. The default is empty string in both cases. If they could be set it could avoid needing a change to the TLC config.

**Comment by emiltin** [comment]
Would that be for testing purposes and what values would be used?

**Comment by otterdahl** [comment]
Yes. For testing purposes. The values needs to be freely configurable, just like the other components "main", "signal_group", etc. in the config file.

**Comment by emiltin** [comment]
ok, separately for each site, like signal groups?

**Comment by otterdahl** [comment]
Yes. For each site, like signal groups

**Comment by emiltin** [comment]
The supervisor should set the two fields when sending messages to the site?

**Comment by otterdahl** [comment]
Yes, when sending, e.g commands. But the site also use them when sending status and alarms to the supervisor.

**Comment by emiltin** [comment]
are ntsOId and xNId configured for each component? or just the site?
must be values be verified when receiving a message?

ref:
from the section https://rsmp-nordic.org/rsmp_specifications/core/3.1.5/applicability/basic_structure.html#basic-structure:

ntsOId | NTSObjectId | Component id for the NTS object which the message is referring to.
xNId | externalNtsId | Identity for the NTS object in communcation between NTS and other systems. The format is 5 integers. Defined in cooperation with representatives from NTS. Unique for the site.
cId | componentId | Component id for the object which the message is referring to.

**Comment by emiltin** [comment]
what's the reason the ntsOId and xNId are empty in most the the alarm messages in https://rsmp-nordic.org/rsmp_specifications/core/3.1.5/applicability/basic_structure.html?highlight=null#structure-for-alarm-request-message?

**Comment by otterdahl** [comment]
> why is ntsOId sent to the site when sending e.g. commands? i think the value is checked during the handshake?

It is because ntsOid is not necessarily unique. It can be differ between components

**Comment by emiltin** [comment]
I still find this confusing. If each component is unique identifies by it's rsmp component id, what do ntsOId refer to and what is the purpose? And how do they related to component id?

**Comment by otterdahl** [comment]
What is ntsOd and xNid?
------------------------
Both ntsOId and xNId are made for the National Traffic Management system (NTS) and refers to an "NTS Object".

What is an "NTS Object"?
------------------------
NTS displays an overview of all sites (in this case traffic lights) and their status. It generally doesn't care about the
status if individual components such as signal groups or detector logics.

This is why "Aggregated status" exists in RSMP. It provides an up-to-date aggregated status of the whole TLC. This also explains
why aggregated status only exists for the main component (or "grouped component" as it is sometimes called).
The main component that provides "aggregated status" is refered to as the "NTS Object". In our case this is the TLC itself.

Why use an additional field "ntsOid"? Why can't we just use component-id?
-------------------------------------------------------------------------
Because NTS also recieves alarms from the TLCs. The alarm may use any component, incl. signal groups and detector logics.
Just by checking at the component-id, NTS doesn't know which TLC it belongs to.

By having ntsOid in the each message (and not just component-id) it is possible for NTS to associate each signal group and detector logic to the indivudal TLC.

This detector alarm example refers to component KK+AG0503=001DL001 in the TLC KK+AG0503=001TC000.

```
{
     "mType":"rSMsg",
     "type":"Alarm",
     "aCId":"A0302",
     "cId":"KK+AG0503=001DL001",
     "ntsOId":"KK+AG0503=001TC000"
}
```

Why not use siteID instead?
---------------------------
Because there may be more than one site and more than one NTS Object in each RSMP connection.

Why use ntsOid and xNId when component-id exists and uniqly identifies a component?
------------------------------------------------------------------------------------
Typically RSMP, NTCIP, OCIT, Modbus, OPC-UA, (or other) is used for communication between the road side equipment and the supervision system. But (at least for us at Trafikverket) the data doesn't stop there. It needs to be sent to NTS (and beyond?)

Between the supervision system and NTS (or other system) other open protocols are typically used (e.g. ASN.1). RSMP usually needs to be translated to this . 

By providing ntsOid and xNId in RSMP it removes the need for the extra step of adding this data in the supervision system by mapping individual components to the main object (NTS Object).

Historically, the need to add additional mapping based on cofiguration on a per site basis has been the source of configuration missmatches. E.g. the configuration of the TLC is updated, but the supervision system is not. Usually this can only be done by different people with different authentication levels.

RSMP were designed to contain all the needed data for NTS already in the TLC and in the same RSMP message, thus removing the need to add this data in the supervision system, before sending this to NTS.

Why is xNId also needed? Why not just ntsOid?
---------------------------------------------
xNId is a diffrent identifier (only integers) also used by NTS. It was added in RSMP for the same reason as meationed above.

What should be sent if NTS doesn't exist?
---------------------------------------------
Leave the fields empty ("")

**Comment by emiltin** [comment]
Thank you for the detailed explanation.

So ntsOId refer to the grouped object which the component belongs to - but instead of using the RSMP identifier, it uses a separat NTS identifier?

This implies that all components belonging to the same grouped object uses the same ntsOId? In that case it would be enough to configure the ntsOId on the grouped object, as long as the configuration specifies which components belong to which grouped object. eg:

```
main component (component id = KK+AG0503=001TC000, ntsOId = NTS-KK+AG0503=001TC000)
  detector logic 1 (KK+AG0503=001DL001)
  detector logic 2 (KK+AG0503=001DL002)
```

In this example, both detectors logics would have ntsOId NTS-KK+AG0503=001TC000 (taken from their parent grouped object).

**Comment by otterdahl** [comment]
> Thank you for the detailed explanation.
> 
> So ntsOId refer to the grouped object which the component belongs to - but instead of using the RSMP identifier, it uses a separat NTS identifier?

Yes. That is correct

> This implies that all components belonging to the same grouped object uses the same ntsOId? In that case it would be enough to configure the ntsOId on the grouped object, as long as the configuration specifies which components belong to which grouped object. eg:
> 
> ```
> main component (component id: KK+AG0503=001TC000, ntsOId: NTS-KK+AG0503=001TC000)
>   detector logic 1 (component id: KK+AG0503=001DL001)
>   detector logic 2 (component id: KK+AG0503=001DL002)
> ```
> 
> In this example, both detectors logics would have ntsOId NTS-KK+AG0503=001TC000 (taken from their parent grouped object).

Yes. This would work.

**Comment by emiltin** [comment]
ok.
is xNId yet another identifer for the same object? i.e. is there a one-to-one relation between ntsOId and xNId (and component id)?

**Comment by otterdahl** [comment]
> ok. is xNId yet another identifer for the same object? i.e. is there a one-to-one relation between ntsOId and xNId (and component id)?

Correct

**Comment by emiltin** [comment]
the current config method in the validator has to be changed. currenty you can set the XNId, but only once per site. it should be set for grouped objects.

**Comment by emiltin** [comment]
something for later discussion - but i think external ids like NTS ids does not belong in RSMP and should be removed in future version. I understand why it was perhaps convenient to add them when RSMP was only used with NTS but for a protocol used with a range of supervisor systems, to me it doesn't seem like a good solution.

**Comment by emiltin** [comment]
can grouped object contained other grouped objects?

**Comment by otterdahl** [comment]
> can grouped object contain other grouped objects? i.e. can they be nested?

No

**Comment by otterdahl** [comment]
> What's the purpose of this, since your NTS system is not involved?

If the NTS system is not used, then there is no need for ntsOid and xNId. In the current version of RSMP core, all fields including ntsOid and xNId are required to exist.

I realize that NTS if very Trafikverket specific. So I'm in favor of making them optional in the next version of RSMP core and not be required unless stated otherwise by the road authoritry.
 
> In effect, it means you need to adress not only the component, but also what grouped object it's in? Shouldn't the component id be enough to uniquely identify it on the site?

Yes, component-id is enough to uniqliy identify it from an RSMP point of view. But as mentioned earlier, there are several benefits of including NTS specifc data in the local equipment (and in the RSMP connection). At least, that was the original reasoning when designing RSMP. First of all, there is no need for additional configuration at the supervision system. Messages (alarms, statuses, commands) can be converted to another protocol without adding ntsOid and xNid. This reduces the chance of configuration missmatches. Second of all, more than one site can be included in a single RSMP connection and there can be several NTS objects per site.
